### PR TITLE
Fix ARM's genPutArgStk codegen

### LINKED ISSUE #23794: Struct arg "reinterpretation" leads to asserts in genPutArgStk on ARM32/64
```C#
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

class Program
{
    [StructLayout(LayoutKind.Sequential)]
    struct S
    {
        public int i0;
        public int i1;
        public int i2;
        public int i3;

        public int i4;
        public int i5;
    }

    [StructLayout(LayoutKind.Sequential)]
    struct S16
    {
        public int i0;
        public int i1;
        public int i2;
        public int i3;
    }

    static void Main() => Test(default);

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Call(int r0, int r1, int r2, int r3, int r4, int r5, int r6, S16 s)
    {
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Escape<T>(ref T t)
    {
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Test(S ss)
    {
        var s = ss;
        Escape(ref s);
        Call(0, 1, 2, 3, 4, 5, 6, Unsafe.As<S, S16>(ref s));
    }
}
```
ARM32:
```
Assert failure(PID 14560 [0x000038e0], Thread: 1428 [0x0594]): Assertion failed '!"Bad value of 'gcType'"' in 'Program:Test(struct)' (IL size 34)

    File: d:\projects\coreclr\src\jit\compiler.cpp Line: 467
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x86.Checked\CoreRun.exe
```
ARM64:
```
Assert failure(PID 1748 [0x000006d4], Thread: 9912 [0x26b8]): Assertion failed 'structSize <= 2 * TARGET_POINTER_SIZE' in 'Program:Test(struct)' (IL size 34)

    File: d:\projects\coreclr\src\jit\codegenarmarch.cpp Line: 786
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x64.Checked\CoreRun.exe
```

### PR DESCRIPTION
**Pull request by mikedn**
When the `OBJ` node wraps a `LCL_VAR` node the code uses the type information (struct size, GC layout) from `LclVarDsc`. This is not always correct because the `OBJ` may actually have a different struct type due to type reinterpretation (e.g. `Unsafe.As<X, Y>`).

Fixes #23794 

No ARM32/64 FX diffs.

### CODE REVIEW

**Comment by mikedn** [comment]
This attempts to be a surgical fix. The existing code is rather confusing and incomplete it seems:
* Attempts to handle lclvars that are not wrapped in a `OBJ` node, this doesn't seem to ever happen.
* Lowering queries GC layout for `OBJ` nodes but it does not call `SetGCInfo` on the `GenTreeObj` node so codegen queries it again (on ARM64). It would seem a no brainer to change lowering to call `SetGCInfo` but doing so may transform the `OBJ` node into a `BLK` node and then codegen does not support that. Well, we could set `GenTreeObj::gtGcPtrs` directly but then perhaps some other code will be surprised to see an `OBJ` with GC info but without actually having GC pointers in it.
* Instead of saving the GC layout information in the `OBJ` node, on ARM32 Lowering saves it in the `GenTreePutArgStk` node. Not quite sure is that, probably to avoid the avoid issue with `OBJ` becoming `BLK`.
* The actual PutArgStk codegen does not support struct sizes that are not a multiple of the slot size. It tries that but only handles cases where the "remaining size" is 1/2/4, not 3 or any other value. I don't know why it even tries that, considering that morph handles such cases by introducing temporary variables. And I don't know why morph does that instead of fixing codegen, call morphing is already complex and fixing codegen is not such a big problem.

Anyway, the existing code is rather convoluted and the only option is a surgical fix or a serious cleanup. The later isn't exactly an option given the coming release...

**Comment by mikedn** [comment]
Test generated code:
```asm
; ARM32
000040  4798           blx     r3               // Program:Escape(byref)
000042  2004           movs    r0, 4
000044  9000           str     r0, [sp] // [V02 OutArgs]
000046  2005           movs    r0, 5
000048  9001           str     r0, [sp+0x04]    // [V02 OutArgs+0x04]
00004A  2006           movs    r0, 6
00004C  9002           str     r0, [sp+0x08]    // [V02 OutArgs+0x08]
00004E  9808           ldr     r0, [sp+0x20]    // [V01 loc0]
000050  9003           str     r0, [sp+0x0c]    // [V02 OutArgs+0x0c]
000052  9809           ldr     r0, [sp+0x24]    // [V01 loc0+0x04]
000054  9004           str     r0, [sp+0x10]    // [V02 OutArgs+0x10]
000056  980A           ldr     r0, [sp+0x28]    // [V01 loc0+0x08]
000058  9005           str     r0, [sp+0x14]    // [V02 OutArgs+0x14]
00005A  980B           ldr     r0, [sp+0x2c]    // [V01 loc0+0x0c]
00005C  9006           str     r0, [sp+0x18]    // [V02 OutArgs+0x18]
00005E  2000           movs    r0, 0
000060  2101           movs    r1, 1
000062  2202           movs    r2, 2
000064  2303           movs    r3, 3
000066  F642 6E78      movw    lr, 0x2e78
00006A  F2C0 7E4E      movt    lr, 0x74e
00006E  F8DE E000      ldr     lr, [lr]
000072  47F0           blx     lr               // Program:Call(int,int,int,int,int,int,int,struct):int

; ARM64
        94000000          bl      Program:Escape(byref)
        A94187A0          ldp     x0, x1, [fp,#24]      // [V01 loc0]
        A90007E0          stp     x0, x1, [sp]  // [V02 OutArgs]
        52800000          mov     w0, #0
        52800021          mov     w1, #1
        52800042          mov     w2, #2
        52800063          mov     w3, #3
        52800084          mov     w4, #4
        528000A5          mov     w5, #5
        528000C6          mov     w6, #6
        94000000          bl      Program:Call(int,int,int,int,int,int,int,struct):int
```

**Comment by BruceForstall** [comment]
cc @dotnet/jit-contrib

File: src/jit/codegenarmarch.cpp [hunk 1]
```diff
@@ -764,12 +764,40 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* treeNode)
...
+                // If the source is an OBJ node then we need to use the type information
+                // it provides (size and GC layout) even if the node wraps a lclvar. Due
+                // to struct reinterpretation (e.g. Unsafe.As<X, Y>) it is possible that
+                // the OBJ node has a different type than the lclvar.
                 CORINFO_CLASS_HANDLE objClass = source->gtObj.gtClass;
 
                 structSize = compiler->info.compCompHnd->getClassSize(objClass);
-                isHfa      = compiler->IsHfa(objClass);
+
+                // The codegen code below doesn't not have proper support for struct sizes
```
Comment (BruceForstall) [inline src/jit/codegenarmarch.cpp]: nit: can you please fix this double negative? "The codegen code below doesn't not have proper support". Maybe just remove the "not"?

Comment (mikedn) [reply src/jit/codegenarmarch.cpp]: Thanks, comment fixed.

**Comment by mikedn** [comment]
Looks like arm crossgen comparison jobs got stuck: "(pending—All nodes of label ‘ubuntu.1404.arm32.open’ are offline) "

**Pull request merged**